apps/opencs/view/render/terrainstorage.cpp

```diff
@@ -33,6 +43,209 @@ namespace CSVRender
+                        if (col == ESM::Land::LAND_SIZE-1 || row == ESM::Land::LAND_SIZE-1)
+                            fixNormal(normal, cellX, cellY, col, row, cache);
+
+                        // some corner normals appear to be complete garbage (z < 0)
+                        if ((row == 0 || row == ESM::Land::LAND_SIZE-1) && (col == 0 || col == ESM::Land::LAND_SIZE-1))
+                            averageNormal(normal, cellX, cellY, col, row, cache);
+
+                        assert(normal.z() > 0);
+
+                        (*normals)[static_cast<unsigned int>(vertX*numVerts + vertY)] = normal;
+
+                        if (colourData)
+                        {
+                            for (int i=0; i<3; ++i)
+                                color[i] = colourData->mColours[srcArrayIndex+i];
+                        }
+                        else
+                        {
+                            color.r() = 255;
+                            color.g() = 255;
+                            color.b() = 255;
+                        }
+
+                        // Highlight broken height changes
+                        if ( ((col > 0 && row > 0) &&
+                            ((abs(heightData->mHeights[col*ESM::Land::LAND_SIZE + row] +
+                            mAlteredHeight[static_cast<unsigned int>(col*ESM::Land::LAND_SIZE + row)] -
+                            (heightData->mHeights[(col)*ESM::Land::LAND_SIZE + row - 1] +
+                            mAlteredHeight[static_cast<unsigned int>((col)*ESM::Land::LAND_SIZE + row - 1)])) >= 1024 ) ||
+                            abs(heightData->mHeights[col*ESM::Land::LAND_SIZE + row] +
+                            mAlteredHeight[static_cast<unsigned int>(col*ESM::Land::LAND_SIZE + row)] -
+                            (heightData->mHeights[(col - 1)*ESM::Land::LAND_SIZE + row] +
+                            mAlteredHeight[static_cast<unsigned int>((col - 1)*ESM::Land::LAND_SIZE + row)]))  >= 1024 )) ||
+                            ((col < ESM::Land::LAND_SIZE - 1 && row < ESM::Land::LAND_SIZE - 1) &&
+                            ((abs(heightData->mHeights[col*ESM::Land::LAND_SIZE + row] +
+                            mAlteredHeight[static_cast<unsigned int>(col*ESM::Land::LAND_SIZE + row)] -
+                            (heightData->mHeights[(col)*ESM::Land::LAND_SIZE + row + 1] +
+                            mAlteredHeight[static_cast<unsigned int>((col)*ESM::Land::LAND_SIZE + row + 1)])) >= 1024 ) ||
+                            abs(heightData->mHeights[col*ESM::Land::LAND_SIZE + row] +
+                            mAlteredHeight[static_cast<unsigned int>(col*ESM::Land::LAND_SIZE + row)] -
+                            (heightData->mHeights[(col + 1)*ESM::Land::LAND_SIZE + row] +
+                            mAlteredHeight[static_cast<unsigned int>((col + 1)*ESM::Land::LAND_SIZE + row)]))  >= 1024 )))
```

Comment:
Please split this expression into multiple statements. Move them to separate functions with meaningful names to explain using what does happen here.

Reply:
Done, did my best in commit bfd7ed1 . Functions require passing around some variables and arrays, which takes some space, but maybe it's still better this way.

Reply:
Looks much better now. It's clear now what is happening there.